### ISSUE TITLE
Nightly build failures, gcc/8.3.0: KokkosCore_UnitTest_DeviceAndThreads: undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#5392]
**Describe the bug**
Nightly cuda/11.2+gcc/8.3.0 on Power9+Volta70 began failing to compile with the `UnitTest_DeviceAndThreads` test:

```
07:08:59 [ 10%] Linking CXX executable KokkosCore_UnitTest_DeviceAndThreads
07:08:59 cd /home/jenkins/weaver-new/workspace/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/Build/core/unit_test && /home/projects/ppc64le/spack-installs/cmake/3.23.1/gcc/7.2.0/base/knlpf2j/bin/cmake -E cmake_link_script CMakeFiles/KokkosCore_UnitTest_DeviceAndThreads.dir/link.txt --verbose=1
07:08:59 /home/jenkins/weaver-new/workspace/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/kokkos/bin/nvcc_wrapper -O3 -DNDEBUG -DKOKKOS_DEPENDENCE -mcpu=power9 -mtune=power9 -arch=sm_70 CMakeFiles/KokkosCore_UnitTest_DeviceAndThreads.dir/UnitTest_DeviceAndThreads.cpp.o -o KokkosCore_UnitTest_DeviceAndThreads  ../src/libkokkoscore.a /usr/lib64/libcuda.so /home/projects/ppc64le-pwr9-nvidia/spack-installs/cuda/11.2.2/gcc/8.3.1/base/uflzqtv/lib64/libcudart.so -ldl 
07:08:59 CMakeFiles/KokkosCore_UnitTest_DeviceAndThreads.dir/UnitTest_DeviceAndThreads.cpp.o: In function `main':
07:08:59 tmpxft_001df942_00000000-6_UnitTest_DeviceAndThreads.cudafe1.cpp:(.text.startup+0x268): undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
07:08:59 collect2: error: ld returned 1 exit status
07:08:59 make[2]: *** [core/unit_test/CMakeFiles/KokkosCore_UnitTest_DeviceAndThreads.dir/build.make:103: core/unit_test/KokkosCore_UnitTest_DeviceAndThreads] Error 1
07:08:59 make[2]: Leaving directory '/home/jenkins/weaver-new/workspace/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/Build'
07:08:59 make[1]: *** [CMakeFiles/Makefile2:2118: core/unit_test/CMakeFiles/KokkosCore_UnitTest_DeviceAndThreads.dir/all] Error 2
07:08:59 make[1]: *** Waiting for unfinished jobs....
```

This failure began following merge of one of these commits:

Add executable that writes num_threads or device_id on demand after initialization ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#5d849927e5289e09d794c4da8b845041d5f5cd79) / [githubweb](https://github.com/kokkos/kokkos/commit/5d849927e5289e09d794c4da8b845041d5f5cd79))
    Add Python test for device_id and num_threads after initialization ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#f213a6aec1bcede024cd77ad7018aae9eb054b39) / [githubweb](https://github.com/kokkos/kokkos/commit/f213a6aec1bcede024cd77ad7018aae9eb054b39))
    Add tests for disable_warnings and tune_internals as well ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#ae3e7fdf5b9bbeb5c2703b332316976b109c6116) / [githubweb](https://github.com/kokkos/kokkos/commit/ae3e7fdf5b9bbeb5c2703b332316976b109c6116))
    Fixup USE_SOURCE_PERMISSIONS is only supported since CMake 3.20 ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#65364d23f51680f54c015c89a00b96e00cbd9709) / [githubweb](https://github.com/kokkos/kokkos/commit/65364d23f51680f54c015c89a00b96e00cbd9709))
    Add OpenMP Target support in the test ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#faad6395dcd635994af104892bc5ea6337102036) / [githubweb](https://github.com/kokkos/kokkos/commit/faad6395dcd635994af104892bc5ea6337102036))
    Add missing <thread> header include ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#c3e1a21af03526a2c3a779097d3414265da22605) / [githubweb](https://github.com/kokkos/kokkos/commit/c3e1a21af03526a2c3a779097d3414265da22605))
    Install newer GCC in ubuntu18.04 based Docker images ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#373e68e4d593d0f9666d0c1865e303e62e1ca6a4) / [githubweb](https://github.com/kokkos/kokkos/commit/373e68e4d593d0f9666d0c1865e303e62e1ca6a4))
    added apt repo ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#a0e3579fb224d6abfa62fea9f8ec3cd1f6618c1c) / [githubweb](https://github.com/kokkos/kokkos/commit/a0e3579fb224d6abfa62fea9f8ec3cd1f6618c1c))
    added header ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#6d4b49b1bd7421fa12f7d65c36c994c8140772c5) / [githubweb](https://github.com/kokkos/kokkos/commit/6d4b49b1bd7421fa12f7d65c36c994c8140772c5))
    pep8ed the python file ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#21807ab584a394a3790aae19b9034e44cc443218) / [githubweb](https://github.com/kokkos/kokkos/commit/21807ab584a394a3790aae19b9034e44cc443218))
    removed DEBIAN_FRONTEND=noninteractive ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#d2d92758fe6c89e96d683a66ea9d8870e98c4004) / [githubweb](https://github.com/kokkos/kokkos/commit/d2d92758fe6c89e96d683a66ea9d8870e98c4004))
    Temporarily disable test for OpenMPTarget since it does not select the right device ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#70962603d6ad722f34084b247bc1aa02e1ca33e4) / [githubweb](https://github.com/kokkos/kokkos/commit/70962603d6ad722f34084b247bc1aa02e1ca33e4))
    Fixup capture_ouptut parameter for subprocess.run was added in version 3.7 ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#93dc5a4e083e45c8f0221fe250f30bcb611e4eb2) / [githubweb](https://github.com/kokkos/kokkos/commit/93dc5a4e083e45c8f0221fe250f30bcb611e4eb2))
    Restrict the number of tests for num_threads ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#867a4ad47aed164fd950c2f7eedf660000d55a4e) / [githubweb](https://github.com/kokkos/kokkos/commit/867a4ad47aed164fd950c2f7eedf660000d55a4e))
    Fixup typo async_arg[c] in ParallelReduce OpenACC ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#045a0d3115d5b5b8290d70680425e1d1263188e0) / [githubweb](https://github.com/kokkos/kokkos/commit/045a0d3115d5b5b8290d70680425e1d1263188e0))
    Implement OpenACC MDRangePolicy parallel_for ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#26945235a98d358c1472094987b9fca2939f2817) / [githubweb](https://github.com/kokkos/kokkos/commit/26945235a98d358c1472094987b9fca2939f2817))
    Enable more tests and fixup unimplemented comments for OpenACC ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#ddddeae355acf07edd102cee442e14715abb7079) / [githubweb](https://github.com/kokkos/kokkos/commit/ddddeae355acf07edd102cee442e14715abb7079))
    Fix initialization for array reductions ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#a9a869957c83147d70efdea7835e109e3302f223) / [githubweb](https://github.com/kokkos/kokkos/commit/a9a869957c83147d70efdea7835e109e3302f223))
    Minor update on the FIXME_OPENACC comment. ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_weaver_cuda_serial_cuda1122_gcc830_volta/11/changes#477df170726c889f14cce3872bc764d521d3808e) / [githubweb](https://github.com/kokkos/kokkos/commit/477df170726c889f14cce3872bc764d521d3808e))

**Please also include the following items to support reproducing the bug**
**Reproducer** (Weaver dev queue):
```bash
module load cuda/11.2.2/gcc/8.3.1 cmake/3.23.1

${KOKKOS_PATH}/generate_makefile.bash --with-cuda --with-serial --compiler=${KOKKOS_PATH}/bin/nvcc_wrapper --arch=Volta70,Power9 --with-cuda-options=enable_lambda --kokkos-path=${KOKKOS_PATH} --cxxflags=\"${CXXFLAGS}\" --cxxstandard=17
```

**Comment by ndellingwood** [comment]
Also failing in gcc/8.3.0 host builds

```
CMakeFiles/KokkosCore_UnitTest_DeviceAndThreads.dir/UnitTest_DeviceAndThreads.cpp.o: In function `std::filesystem::__cxx11::path::path<char*, std::filesystem::__cxx11::path>(char* const&, std::filesystem::__cxx11::path::format)':
/net/watson.sandia.gov/storage/fast/projects/sems/install/rhel7-x86_64/sems/compiler/gcc/8.3.0/base/include/c++/8.3.0/bits/fs_path.h:184: undefined reference to `std::filesystem::__cxx11::path::_M_split_cmpts()'
collect2: error: ld returned 1 exit status
```

**Comment by ndellingwood** [comment]
Failing with the Hip backend as well, rocm/4.5.0+gcc/8

**Comment by ndellingwood** [comment]
Same error with intel/19 builds as well

**Comment by masterleinad** [comment]
Looks like we need `-lstdc++fs` for some compilers.

**Comment by ndellingwood** [comment]
@masterleinad It doesn't look like there were issues with gcc/9.2 or gcc/10.2 builds, apparently specific to gcc/8.* (or older)

**Comment by masterleinad** [comment]
https://en.cppreference.com/w/cpp/filesystem#Notes says:
```
Using this library may require additional compiler/linker options. GNU implementation prior to 9.1 requires linking with -lstdc++fs and LLVM implementation prior to LLVM 9.0 requires linking with -lc++fs. 
```

**Comment by ndellingwood** [comment]
Another problem with the `KokkosCore_UnitTest_DeviceAndThreads` test when attempting to configure Trilinos (cuda/11.1+gcc/9.2.0):

```
-- Configuring done
CMake Error at kokkos/core/unit_test/CMakeLists.txt:1066 (add_executable):
  Target "KokkosCore_UnitTest_DeviceAndThreads" links to target
  "Kokkos::kokkoscore" but the target was not found.  Perhaps a
  find_package() call is missing for an IMPORTED target, or an ALIAS target
  is missing?


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Does something require the "Dummy.cpp" type treatment that was needed for the recent simd related fixes?

**Comment by dalg24** [comment]
> Another problem with the `KokkosCore_UnitTest_DeviceAndThreads` test when attempting to configure Trilinos (cuda/11.1+gcc/9.2.0):
> 
> ```
> -- Configuring done
> CMake Error at kokkos/core/unit_test/CMakeLists.txt:1066 (add_executable):
>   Target "KokkosCore_UnitTest_DeviceAndThreads" links to target
>   "Kokkos::kokkoscore" but the target was not found.  Perhaps a
>   find_package() call is missing for an IMPORTED target, or an ALIAS target
>   is missing?
> 
> 
> -- Generating done
> CMake Generate step failed.  Build files cannot be regenerated correctly.
> ```
> 
> Does something require the "Dummy.cpp" type treatment that was needed for the recent simd related fixes?

Two possible resolution:
1. (my strong preference) disable the test when built as part of Trilinos
```CMake
if(NOT KOKKOS_HAS_TRILINOS)
  ...
endif()
```
3. Refactor
https://github.com/kokkos/kokkos/blob/aef15d2e0a28511eea6695a02ab3bfc182a723ba/core/unit_test/CMakeLists.txt#L1066-L1067
to use `kokkos_add_executable` (likely will need more adjustments)

**Comment by dalg24** [comment]
Or am I reading it wrong?  Was it indeed with Trilinos?

**Comment by ndellingwood** [comment]
> Was it indeed with Trilinos?

@dalg24 yes, this occurred when configuring Trilinos with kokkos' develop branch

**Comment by masterleinad** [comment]
Fixed in #5402 and #5404 (and only relevant for `develop`).
@ndellingwood Please reopen if the problem persists for whatever reason.
